### PR TITLE
Test demo as refered on https://vladmihalcea.com/how-to-implement-equ…

### PR DESCRIPTION
**Pull request by JosRoseboom**
Test demo as referred on https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier/ : suppose someone creates a DAO with a general way of putting both transient and detached objects in managed state, implementing it with entityManager.merge. Then I guess your test would fail because in the equals method it is not the same object (since merge returns a new object) and the id is not the same (since the transient object does not have one). Am i correct that this goes wrong? And is the solution just to never use such a general way of putting objects in managed state?

### CODE REVIEW

**Comment by vladmihalcea** [comment]
Merge is not a generic save. It's for detached entities only. If you are talking about cascading merge, then you use [this test case](https://github.com/vladmihalcea/high-performance-java-persistence/blob/11966856f4b1972cd98da06c2a0b7f8cd4d0bf1e/core/src/test/java/com/vladmihalcea/book/hpjp/hibernate/association/BidirectionalOneToManyMergeTest.java) to prove the issue.

For the moment, this test case shows an improper JPA usage. Calling merge instead of persist for transient entities is a mistake.

**Comment by JosRoseboom** [comment]
Thanks, that is the answer to my question. Not using merge for transient objects seems a proper convention.

However, how could I (or anybody in general) know this? The jpa docs on https://docs.oracle.com/javaee/7/api/javax/persistence/EntityManager.html says on merge "Merge the state of the given entity into the current persistence context." The hibernate docs on https://docs.jboss.org/hibernate/entitymanager/3.6/reference/en/html_single/ only indicate in 3.8 how it is usually used, not how it should be used. And, most of all, the 'Java persistence with hibernate' book says on page 252 (10.3.4) of the second edition:

_If the instance you’re giving to merge() is not detached but rather is transient (it doesn’t have an identifier value), Hibernate instantiates a fresh User, copies the values of the transient User onto it, and then makes it persistent and returns it to you. In simpler terms, the merge() operation can handle detached and transient entity instances. Hibernate always returns the result to you as a persistent instance._

_An application architecture based on detachment and merging may not call the persist() operation. You can merge new and detached entity instances to store data. The important difference is the returned current state and how you handle this switch of references in your application code. You have to discard the detachedUser and from now on reference the current mergedUser. Every other component in your application still holding on to detachedUser has to switch to mergedUser._

Is not using merge for transient objects opinionated or did I miss something?

**Comment by vladmihalcea** [comment]
The Hibernate Docs you linked are very old. When I rewrote the Hibernate User Guide, I explained how `merge` works. Check out [merging detached data section in the new User Guide](https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#pc-merge) for more details.

> Is not using merge for transient objects opinionated or did I miss something?

No, it's not. If you call `merge` for 100 entities that use the `assigned` identifier generator, you will end up with 100 extra SELECT statements being executed. That's not very efficient.

In your example, the UI knows whether the client added new entities, updated existing ones, or deleted any of the existing ones. 

For the new ones, you should call `persist`, for the existing ones, you could call `update` or `merge`.

However, if you're talking about child entities that have a collection association on the parent side, then `merge` could work for new/changed/removed child entries because Hibernate knows the loading-time state and propagates the changes accordingly. Check out [this article](https://vladmihalcea.com/merge-entity-collections-jpa-hibernate/) for more details.
